### ISSUE TITLE
Image test demo

### ISSUE THREAD

**Issue by Cronusf** [GengDavid/pytorch-cpn#20]
How to use the pre-training network model to visualize the image test results? Is there a demo?

**Comment by YinRui1991** [comment]
@GengDavid Thanks for your pytorch re-implementation of CPN.
@Cronusf Hi, I write a simple single image process demo, you can modify this to test images or videos. Maybe this can help you.

import os
import sys
import argparse
import torch
import torch.nn.parallel
import torch.optim
import cv2
import numpy as np

from test_config import cfg
from networks import network
import time
from utils.transforms import *

def augmentationCropImage(cfg, img, bbox):

    height, width = cfg.data_shape[0], cfg.data_shape[1]
    bbox = np.array(bbox).reshape(4, ).astype(np.float32)
    add = max(img.shape[0], img.shape[1])
    mean_value = cfg.pixel_means
    bimg = cv2.copyMakeBorder(img, add, add, add, add, borderType=cv2.BORDER_CONSTANT, value=mean_value.tolist())
    objcenter = np.array([(bbox[0] + bbox[2]) / 2., (bbox[1] + bbox[3]) / 2.])
    bbox += add
    objcenter += add

    crop_width = (bbox[2] - bbox[0]) * (1 + cfg.bbox_extend_factor[0] * 2)
    crop_height = (bbox[3] - bbox[1]) * (1 + cfg.bbox_extend_factor[1] * 2)

    if crop_height / height > crop_width / width:
        crop_size = crop_height
        min_shape = height
    else:
        crop_size = crop_width
        min_shape = width

    crop_size = min(crop_size, objcenter[0] / width * min_shape * 2. - 1.)
    crop_size = min(crop_size, (bimg.shape[1] - objcenter[0]) / width * min_shape * 2. - 1)
    crop_size = min(crop_size, objcenter[1] / height * min_shape * 2. - 1.)
    crop_size = min(crop_size, (bimg.shape[0] - objcenter[1]) / height * min_shape * 2. - 1)

    min_x = int(objcenter[0] - crop_size / 2. / min_shape * width)
    max_x = int(objcenter[0] + crop_size / 2. / min_shape * width)
    min_y = int(objcenter[1] - crop_size / 2. / min_shape * height)
    max_y = int(objcenter[1] + crop_size / 2. / min_shape * height)

    img = cv2.resize(bimg[min_y:max_y, min_x:max_x, :], (width, height))
    details = np.asarray([min_x - add, min_y - add, max_x - add, max_y - add]).astype(np.float)

    return img, details

def load_image(img_path,gt_bbox,cfg):

    image = scipy.misc.imread(img_path, mode='RGB')  # image is a numpy image
    image, details = augmentationCropImage(cfg, image, gt_bbox)

    img = im_to_torch(image)
    img = color_normalize(img, cfg.pixel_means)

    return img, details

def main(args):

    # create model
    model = network.__dict__[cfg.model](cfg.output_shape, cfg.num_class, pretrained = False)
    model = torch.nn.DataParallel(model, device_ids=[0]).cuda()

    # load trainning weights
    checkpoint_file = os.path.join(args.checkpoint, args.test+'.pth.tar')
    checkpoint = torch.load(checkpoint_file)
    model.load_state_dict(checkpoint['state_dict'])
    print("=> loaded checkpoint '{}' (epoch {})".format(checkpoint_file, checkpoint['epoch']))
    
    # change to evaluation mode
    model.eval()
    
    print('testing...')
    full_result = []
    t1 = time.time()
    image_path = '1.jpg'
    im = cv2.imread(image_path)
    bboxes = (5,5,159,424)  # bbox (x1,y1,x2,y2)
    inputs,details = load_image(image_path, bboxes, cfg)
    inputs = inputs.unsqueeze(0)

    with torch.no_grad():
        input_var = torch.autograd.Variable(inputs.cuda())

        if args.flip == True:
            flip_inputs = inputs.clone()
            for i, finp in enumerate(flip_inputs):
                finp = im_to_numpy(finp)
                finp = cv2.flip(finp, 1)
                flip_inputs[i] = im_to_torch(finp)
            flip_input_var = torch.autograd.Variable(flip_inputs.cuda())

        # compute output
        global_outputs, refine_output = model(input_var)
        score_map = refine_output.data.cpu()
        score_map = score_map.numpy()

        if args.flip == True:
            flip_global_outputs, flip_output = model(flip_input_var)
            flip_score_map = flip_output.data.cpu()
            flip_score_map = flip_score_map.numpy()

            for i, fscore in enumerate(flip_score_map):
                fscore = fscore.transpose((1,2,0))
                fscore = cv2.flip(fscore, 1)
                fscore = list(fscore.transpose((2,0,1)))
                for (q, w) in cfg.symmetry:
                   fscore[q], fscore[w] = fscore[w], fscore[q]
                fscore = np.array(fscore)
                score_map[i] += fscore
                score_map[i] /= 2

        for b in range(inputs.size(0)):

            single_result_dict = {}
            single_result = []

            single_map = score_map[b]
            r0 = single_map.copy()
            r0 /= 255
            r0 += 0.5
            v_score = np.zeros(17)
            for p in range(17):
                single_map[p] /= np.amax(single_map[p])
                border = 10
                dr = np.zeros((cfg.output_shape[0] + 2*border, cfg.output_shape[1]+2*border))
                dr[border:-border, border:-border] = single_map[p].copy()
                dr = cv2.GaussianBlur(dr, (21, 21), 0)
                lb = dr.argmax()
                y, x = np.unravel_index(lb, dr.shape)
                dr[y, x] = 0
                lb = dr.argmax()
                py, px = np.unravel_index(lb, dr.shape)
                y -= border
                x -= border
                py -= border + y
                px -= border + x
                ln = (px ** 2 + py ** 2) ** 0.5
                delta = 0.25
                if ln > 1e-3:
                    x += delta * px / ln
                    y += delta * py / ln
                x = max(0, min(x, cfg.output_shape[1] - 1))
                y = max(0, min(y, cfg.output_shape[0] - 1))
                #resy = float((4 * y + 2) / cfg.data_shape[0] * (details[b][3] - details[b][1]) + details[b][1])
                #resx = float((4 * x + 2) / cfg.data_shape[1] * (details[b][2] - details[b][0]) + details[b][0])
                resy = float((4 * y + 2) / cfg.data_shape[0] * (details[3] - details[1]) + details[1])
                resx = float((4 * x + 2) / cfg.data_shape[1] * (details[2] - details[0]) + details[0])
                v_score[p] = float(r0[p, int(round(y)+1e-10), int(round(x)+1e-10)])
                single_result.append(resx)
                single_result.append(resy)
            if len(single_result) != 0:
                single_result_dict['keypoints'] = single_result
                print (single_result)
                for k in range(17):
                    x, y = int(single_result[k*2]), int(single_result[k*2+1])
                    cv2.circle(im, (x, y), 2, (255, 0, 0), 2)
                cv2.imwrite('res.jpg', im)
                full_result.append(single_result_dict)

    print (time.time()-t1)


if __name__ == '__main__':

    parser = argparse.ArgumentParser(description='PyTorch CPN Test')
    parser.add_argument('-j', '--workers', default=12, type=int, metavar='N',
                        help='number of data loading workers (default: 12)')
    parser.add_argument('-g', '--num_gpus', default=1, type=int, metavar='N',
                        help='number of GPU to use (default: 1)')      
    parser.add_argument('-c', '--checkpoint', default='checkpoint', type=str, metavar='PATH',
                        help='path to load checkpoint (default: checkpoint)')
    parser.add_argument('-f', '--flip', default=True, type=bool,
                        help='flip input image during test (default: True)')
    parser.add_argument('-b', '--batch', default=128, type=int,
                        help='test batch size (default: 128)')
    parser.add_argument('-t', '--test', default='CPN256x192', type=str,
                        help='using which checkpoint to be tested (default: CPN256x192')
    parser.add_argument('-r', '--result', default='result', type=str,
                        help='path to save save result (default: result)')
    main(parser.parse_args())

You can create a demo_im.py file at ROOT_DIR/MODEL_DIR/, 
and run : python3 demo_im.py  -t 'CPN256x192'

In this demo code, the bbox is given by hand. You can add a person detector to generate bboxes, then you can process images or videos.

**Comment by binhmuc** [comment]
THe same question, how to test with an image for demo

**Comment by ZhuoerLyu** [comment]
Did you guys figure it out? I also have the same question.

**Comment by YinRui1991** [comment]
Refer the test file and add a detector or give bbox by hand.

**Comment by linhaibo123** [comment]
test with an image, you must provide image and bbox,so need to add a detection network in front, or give some fixed bboxes

**Comment by chh7411898** [comment]
THe same question, how to test with an image for demo
